visualization/app/codeCharta/util/treeMapGenerator.ts

```diff
@@ -1,44 +1,108 @@
+		this.scaleRoot(nodes[0], scale)
 
-	private static getNodesAsArray(node: HierarchyRectangularNode<CodeMapNode>): HierarchyRectangularNode<CodeMapNode>[] {
-		const nodes = [node]
-		if (node.children) {
-			node.children.forEach(child => nodes.push(...this.getNodesAsArray(child)))
+		for (const fixedFolder of map.children) {
+			const squarified = this.getSquarifiedTreeMap(fixedFolder, state)
+			squarified.treeMap.descendants().forEach(squarifiedNode => {
+				this.scaleAndTranslateSquarifiedNode(squarifiedNode, fixedFolder, squarified, scale)
+				const node = TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, state, isDeltaState)
+				nodes.push(node)
+			})
 		}
 		return nodes
 	}
 
+	private static scaleAndTranslateSquarifiedNode(
+		squarifiedNode: HierarchyRectangularNode<CodeMapNode>,
+		fixedFolder: CodeMapNode,
+		squarified: SquarifiedTreeMap,
+		scale: number
+	) {
+		// Transform coordinates from local folder space to world space (between 0 and 100).
+		const scaleX = fixedFolder.fixed.width / squarified.width
+		const scaleY = fixedFolder.fixed.height / squarified.height
+
+		// Scales to usual map-size of 500 matching the three-scene-size
+		squarifiedNode.x0 = (squarifiedNode.x0 * scaleX + fixedFolder.fixed.x) * scale
+		squarifiedNode.x1 = (squarifiedNode.x1 * scaleX + fixedFolder.fixed.x) * scale
+		squarifiedNode.y0 = (squarifiedNode.y0 * scaleY + fixedFolder.fixed.y) * scale
+		squarifiedNode.y1 = (squarifiedNode.y1 * scaleY + fixedFolder.fixed.y) * scale
+	}
+
+	private static scaleRoot(root: Node, scale: number) {
+		root.x0 *= scale
+		root.y0 *= scale
+		root.width *= scale
+		root.length *= scale
+	}
+
+	private static getSquarifiedTreeMap(map: CodeMapNode, s: State): SquarifiedTreeMap {
+		const hierarchyNode: HierarchyNode<CodeMapNode> = hierarchy<CodeMapNode>(map)
+		const nodesPerSide = this.getEstimatedNodesPerSide(hierarchyNode)
+		const padding: number = s.dynamicSettings.margin * TreeMapGenerator.PADDING_SCALING_FACTOR
+		let mapWidth
+		let mapHeight
+
+		if (map.fixed !== undefined) {
+			mapWidth = map.fixed.width
+			mapHeight = map.fixed.height
+		} else {
+			mapWidth = s.treeMap.mapSize * 2
+			mapHeight = s.treeMap.mapSize * 2
+		}
+
+		const width = mapWidth + nodesPerSide * s.dynamicSettings.margin
+		const height = mapHeight + nodesPerSide * s.dynamicSettings.margin
+
+		const treeMap: TreemapLayout<CodeMapNode> = treemap<CodeMapNode>().size([width, height]).paddingOuter(padding).paddingInner(padding)
+
+		return { treeMap: treeMap(hierarchyNode.sum(node => this.calculateAreaValue(node, s))), height, width }
+	}
+
+	private static getEstimatedNodesPerSide(hierarchyNode: HierarchyNode<CodeMapNode>): number {
+		const descendants = hierarchyNode.descendants()
+		const blacklisted = descendants.filter(node => node.data.isExcluded || node.data.isFlattened).length
+		return 2 * Math.sqrt(descendants.length - blacklisted)
```

Comment:
How come it's multiplied by 2? :-) Maybe add a comment?

Reply:
I wish I could explain this. This piece of code has been there since I started. All I know is, that it somehow estimates the number of nodes that are on each side, so I moved it to it's own function to explain at least this.

Reply:
I see, thanks!